src/BSON/Iterator.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "BSON/Iterator.h"
+#include "BSON/Iterator_arginfo.h"
+
+zend_class_entry* php_phongo_iterator_ce;
+
+static bool php_phongo_iterator_init_with_zval(php_phongo_iterator_t* iterator, zval* zbson)
+{
+	php_phongo_document_t*  document_intern;
+	php_phongo_arraylist_t* arraylist_intern;
+	const bson_t*           bson;
+
+	ZVAL_COPY(&iterator->bson, zbson);
+	if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_document_ce)) {
+		document_intern = Z_DOCUMENT_OBJ_P(&iterator->bson);
+		bson            = document_intern->bson;
+	} else if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_arraylist_ce)) {
+		arraylist_intern   = Z_ARRAYLIST_OBJ_P(&iterator->bson);
+		bson               = arraylist_intern->bson;
+		iterator->is_array = true;
+	} else {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Could not create iterator for %s instance.", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zbson));
```

Comment:
For consistency, we don't use periods for single-sentence exception messages.

Also, since this function is only called internally, I think this particular case could be a logic excepiton like we have in `phongo_execute.c`:

```
/* Should never happen, but if it does: exception */
phongo_throw_exception(PHONGO_ERROR_LOGIC, "Type '%d' should never have been passed to phongo_execute_command, please file a bug report", type);
```

Reply:
Changed error type to `PHONGO_ERROR_LOGIC`, removed periods at the end. I did not add the "please file a bug report" part.